### ISSUE TITLE
[Bug] Cache seen-token count is committed too early during decode, shifting upper-layer RoPE offsets

### ISSUE THREAD

**Issue by vpuri3** [fla-org/flash-linear-attention#766]
### Checklist

- [ ] I have checked [FAQs](https://github.com/fla-org/flash-linear-attention/blob/main/FAQs.md) and existing issues for similar problems
- [x] Please report this bug in English to ensure wider understanding and support

## Summary

The shared cache implementation updates the global seen-token count too early during cached decoding. When `layer 0` writes its new KV state for the current token, `_seen_tokens` is incremented immediately, so upper layers in the same forward observe a past length that is already one token too large.

This makes `past_key_values.get_seq_length(layer_idx)` return `past_len + 1` for upper layers on the current decode step, which shifts RoPE positions by one token and breaks exact cache equivalence.

## Affected code

The issue is in the shared cache classes in `fla/models/utils.py`:

- `LegacyFLACache.update()` increments `_seen_tokens` when `layer_idx == 0` (or at the last existing layer in the legacy branch) before the rest of the model finishes the same forward.
- `FLACache.update()` also increments `_seen_tokens` immediately when `layer_idx == 0`.

Attention layers then use `past_key_values.get_seq_length(self.layer_idx)` to build `seqlen_offset` for RoPE, so upper layers read the inflated count on the same step.

## Minimal mechanism

On decode step `t`:

1. Layer 0 runs and appends the new token to cache.
2. `_seen_tokens` is incremented from `t` to `t+1` immediately.
3. Layer 1 (and all higher layers) call `get_seq_length()` during the same forward.
4. They receive `t+1` instead of the correct past length `t` and apply RoPE with an off-by-one offset.

## Suggested fix

Defer the global seen-token update until the next layer-0 read (or otherwise until the current forward is complete), e.g. by storing a pending token count during `update()` and only folding it into `_seen_tokens` when a new decode/prefill step begins.

The key invariant is that all layers in one forward pass must observe the same past length.

## Notes

This is a shared cache bug and not specific to `flare`. It affects any model path that relies on `get_seq_length()` during decode for positional offsets.

### Steps to Reproduce the Bug

• The original failing test was tests/models/test_modeling_bitnet.py:48, specifically:

  - test_generation[L2-B4-T2000-H8-D64-torch.float16]

  That test calls the shared helper run_test_generation:69, which is where the cache-vs-reference comparison logic
  lives.

  I also confirmed the same shared issue showed up in tests/models/test_modeling_transformer.py:48 on its matching:

  - test_generation[L2-B4-T2000-H8-D64-torch.float16]

  So the most relevant place to inspect is really the shared harness:

  - tests/models/test_modeling_base.py:69

### Expected Behavior

This showed up as cache-based generation mismatches in multiple models. In particular, I reproduced it in the shared `test_generation` harness with at least:

- `Transformer`
- `BitNet`

Before the fix, decode checkpoints differed from a fresh cached full-prefix forward; after deferring the seen-token commit, those checkpoint logits matched exactly.

### Environment Information

1. Torch: 2.8cu128
2. Triton:

**Comment by zhiyuan1i** [comment]
Hi @vpuri3,

Thank you for reporting this issue! After investigating, I'm happy to report that **this bug has already been fixed** in the codebase.

## Fix Details

**Fix Commit:** `decd1978` ([Cache] Fix get_seq_length to return per-layer length, #748)  
**Date:** 2026-02-14  
**Author:** Zhiyuan Li

### Root Cause
The issue was that `FLACache` used a global `_seen_tokens` counter shared across all layers. When Layer 0 updated its cache, the global counter incremented immediately, causing upper layers to see an inflated past length during the same forward pass.

### Solution
The fix introduced per-layer `_seen_tokens` tracking via `FLALayer`:
- Each `FLALayer` now maintains its own independent `_seen_tokens` counter
- `FLACache.get_seq_length(layer_idx)` delegates to the corresponding layer's counter
- Layer updates no longer affect other layers' sequence length calculations

## Verification

I've added comprehensive regression tests in `tests/models/test_cache.py` to ensure this issue doesn't recur:

```bash
$ pytest tests/models/test_cache.py -v
============================= test session starts ==============================
...
tests/models/test_cache.py::test_cache_decode_all_layers_see_same_past_length PASSED
tests/models/test_cache.py::test_all_layers_see_consistent_past_len_during_decode_regression PASSED
tests/models/test_cache.py::test_rope_offset_consistency_simulation_regression PASSED
...
================== 21 passed, 2 skipped in 0.32s ===============================
```

All tests pass, confirming that:
1. All layers see consistent `past_len` during decode
2. RoPE offsets are calculated correctly across all layers
3. Per-layer counters work independently

## Recommendation

This issue can now be **closed**. If you still encounter cache-related generation mismatches on the latest main branch, please let us know with specific reproduction steps.

Thanks again for the detailed bug report!
